wallet: Only fail rescan when blocks have actually been pruned

maflcko opened the pull request:
This brings the behaviour of the import* calls closer to importmulti. After this change, the difference between importmulti and the other import* calls is

* that in importmulti you can "opt-out" of scanning early blocks by setting a later timestamp.
* that in importmulti the wallet will successfully import the data, but fail to rescan. Whereas in the other calls, the wallet will abort before importing the data.

Commit:
wallet: Only fail rescan when blocks have actually been pruned

gmaxwell commented:
Concept ACK.

There is an argument against doing things like this:  that it's better to fail deterministically rather than fail randomly at some point in the future and have created incorrect expectations of functionality.  But setting prune to some large "future" amount is a perfectly reasonable configuration.  And this doesn't cause it to fail randomly, but rather work until it doesn't.  So I don't believe the general argument against this sort of thing offsets the benefit.

ryanofsky commented:
This seems to affect importaddress, importpubkey, importprivkey, and importwallet but not importmulti. I wonder how the new behavior compares to importmulti behavior, and if there's a difference, whether that's justified.

MarcoFalke commented:
@ryanofsky The only difference between importmulti and the other calls is that in importmulti you can "opt-out" of scanning early blocks by setting a later timestamp. Otherwise, my change makes all calls behave in the same way.

MarcoFalke commented:
Oh, my previous reply is untrue...

MarcoFalke commented:
I believe the only difference is that importmulti will import the keys and then fail the rescan, whereas the other calls will exit early and not attempt to import anything.

MarcoFalke commented:
Updated OP to say that

promag commented:
> Edit: I believe the only difference is that importmulti will import the keys and then fail the rescan, whereas the other calls will exit early and not attempt to import anything.

Why a different behavior? It's already possible to call `importmulti ... '{"rescan": false}'` and then `rescanblockchain ...`? I think RPCs should (try to) be atomic so that the client doesn't have to figure out what succeeded or not.

MarcoFalke commented:
> Why a different behavior?

I don't know. I didn't write the importmulti RPC and I don't plan to change it in this pull request.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -326,7 +326,11 @@ class ChainImpl : public Chain
     CFeeRate relayDustFee() override { return ::dustRelayFee; }
     CAmount maxTxFee() override { return ::maxTxFee; }
-    bool getPruneMode() override { return ::fPruneMode; }
+    bool havePruned() override
```
Move to the chain lock?

MarcoFalke replied:
Why?

MarcoFalke replied:
Sorry for the brief reply, but see here:

https://github.com/bitcoin/bitcoin/blob/fad84519391a05f2ad42381cca9135323b709657/src/interfaces/chain.h#L69-L73

The chain lock is deprecated and will be removed in the future.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -157,8 +157,8 @@ UniValue importprivkey(const JSONRPCRequest& request)
 
-        if (fRescan && pwallet->chain().getPruneMode()) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "Rescan is disabled in pruned mode");
+        if (fRescan && pwallet->chain().havePruned()) {
```
A block can be pruned before RescanWallet (L202). What would be the error then?

MarcoFalke replied:
It will throw an error `"Rescan was unable to fully rescan the blockchain. Some transactions may be missing."`

promag reviewed (commented):
@MarcoFalke I wasn't suggesting changing that behavior.

Concept ACK. Correct me if I'm wrong, but it doesn't have to fail if all required blocks are available right?

MarcoFalke commented:
> Concept ACK. Correct me if I'm wrong, but it doesn't have to fail if all required blocks are available right?

importmulti does not fail the rescan when all blocks after the timestamp are available, but it fails *after* the import if a block is missing.
the other import* (after my change) also do not fail the rescan when all blocks after time=0 are available. However, they fail *before* the import if a block is missing.

luke-jr commented:
Let's use prune height instead, so `rescanblockchain` can be picky about it?

MarcoFalke commented:
I am not touching `rescanblockchain` at all. Could you explain what you mean?

luke-jr commented:
I mean it would make sense to use a common interface with `rescanblockchain`, which would need prune-height rather than have-pruned.

MarcoFalke commented:
Oh, the RPCs I touch are already deprecated in favor of importmulti, so I'd rather not add features to them.

luke-jr commented:
I'm commenting on the src/interfaces changes specifically.

MarcoFalke commented:
Ah. Yeah, since we don't store the prune height, this would complicate my changes slightly. I think the additional code can be added later (when needed) in rescanblockchain.

jnewbery commented:
I'm a little concerned about this change. Anywhere that `havePruned()` is called without `cs_main` is racy, since a block may be pruned after `havePruned()` is called. We therefore need to make sure that the code after those `havePruned()` calls is robust against being run when blocks have been pruned. If that's the case, why not just remove the `chain().getPruneMode()` check entirely and let the error-handling in the later code handle when a block is missing?

The four call sites in rpcdump do not hold `cs_main`. The call site in `CreateWalletFromFile()` does hold `cs_main` indirectly through `locked_chain`, but a goal is to remove that lock.

MarcoFalke commented:
@jnewbery The check is only there to provide a nice error message. As you correctly observe it is not needed for correctness. Should I clarify that with a comment?

The error message otherwise would be https://github.com/bitcoin/bitcoin/pull/15870#discussion_r278315111

jnewbery commented:
> The check is only there to provide a nice error message. As you correctly observe it is not needed for correctness. Should I clarify that with a comment?

Yes, that sounds good. Thanks!

jnewbery reviewed (commented):
utACK fa9fa5d647092710aa03dadf92433584e9fe26a3

Will happily reACK a squashed branch.

MarcoFalke commented on test/functional/wallet_import_rescan.py:
```diff
@@ -40,9 +39,6 @@ class Variant(collections.namedtuple("Variant", "call data rescan prune")):
-        if self.expect_disabled:
-            assert_raises_rpc_error(-4, "Rescan is disabled in pruned mode", func, *args, **kwargs)
-        else:
             return func(*args, **kwargs)
```
Done

jnewbery commented:
utACK fa9f3cc89442e9b332a27b81b04201bb4a0caf28

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -157,8 +157,11 @@ UniValue importprivkey(const JSONRPCRequest& request)
         if (!request.params[2].isNull())
```
Nit: Update year in line 1? `// Copyright (c) 2009-2018 The Bitcoin Core developers`

MarcoFalke replied:
Done with a scripted diff

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -4221,8 +4221,11 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     {
         //We can't rescan beyond non-pruned blocks, stop and throw an error
         //this might happen if a user uses an old wallet within a pruned node
```
Nit: space after // in lines 4222 and 4223

jonatack replied:
Perhaps out of PR scope but the docs in lines 4222 and 4223 could be clearer, e.g. is "error" the end of a sentence and "this" the start of a sentence, or...

MarcoFalke replied:
This is only a dev comment, so no need to over-optimize. Will take specific suggestions if I need to touch this pull for other reasons, though.

jnewbery replied:
I don't think "no need to over-optimize" is a helpful response. The grammar is wrong. Why not correct it?

MarcoFalke replied:
Added a dot and space if that is what you meant

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -573,7 +582,10 @@ UniValue importwallet(const JSONRPCRequest& request)
+        // Exit early and print an error.
+        // If a block is pruned after this check, we will import the key(s),
+        // but fail the rescan with a generic error.
         throw JSONRPCError(RPC_WALLET_ERROR, "Importing wallets is disabled in pruned mode");
```
Does the code doc in 588 correspond correctly with the error message in 589?

MarcoFalke replied:
Yes, "importing wallet" means "importing key(s)", I believe.

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -314,8 +317,11 @@ UniValue importaddress(const JSONRPCRequest& request)
+        // Exit early and print an error.
+        // If a block is pruned after this check, we will import the key(s),
+        // but fail the rescan with a generic error.
+        throw JSONRPCError(RPC_WALLET_ERROR, "Rescan is disabled when blocks are pruned");
```
Would it be worth extracting this code block repeated in `importprivkey`, `importaddress`, `importpubkey`, and `importwallet` (the latter with a different error message).

```cpp
    if (fRescan && pwallet->chain().havePruned()) {
        // Exit early and print an error.
        // If a block is pruned after this check, we will import the key(s),
        // but fail the rescan with a generic error.
        throw JSONRPCError(RPC_WALLET_ERROR, "Rescan is disabled when blocks are pruned");
    }
```

MarcoFalke replied:
Yeah, could make sense. Will do if I have to touch for other reasons.

promag replied:
faf3729242c5e9486e137e549f05bf20bd3908a0

Yeah, in a follow up is fine.

jonatack reviewed (commented):
A few code comments. More documentation and tests might make this change and the expected behavior before/after clearer, at least to me.

Commit:
scripted-diff: Bump copyright headers in wallet

-BEGIN VERIFY SCRIPT-
./contrib/devtools/copyright_header.py update ./src/wallet/
-END VERIFY SCRIPT-

sipa commented:
Concept ACK. My first read of this was that it'd permit rescanning pruned chains, as long as the to-be-rescanned blocks are not pruned, but it seems to just be about whether or not any blocks are actually pruned (independent of the ones to be rescanned), right?

MarcoFalke commented:
The legacy calls would always rescan the whole chain. As soon as one block is pruned, it fails.

wpaulino commented:
Out of scope for this PR but somewhat related -- would there be any reason to not attempt to fetch the pruned blocks from the P2P network to allow the rescan to happen?

MarcoFalke commented:
Jup, but that might be something to do only after #15946

ryanofsky commented on test/functional/wallet_import_rescan.py:
```diff
@@ -39,23 +38,17 @@
 
-    def try_rpc(self, func, *args, **kwargs):
-        if self.expect_disabled:
-            assert_raises_rpc_error(-4, "Rescan is disabled in pruned mode", func, *args, **kwargs)
```
Now there is no longer any test coverage for the "rescan is disabled" cases, but I'm not sure if there's an easy way to add it back. Maybe a `pruneblockchain` RPC call could be added.

MarcoFalke replied:
If a test was added, it should also (or more importantly) test `importmulti`, so I am not actually decreasing test coverage here.

MarcoFalke replied:
I agree that a test would be nice for that case, but can be done in a follow up pull request.

ryanofsky replied:
> it should also (or more importantly) test importmulti

In case it helps, there is currently a unit test for importmulti behavior with pruning: https://github.com/bitcoin/bitcoin/blob/2d16fb7a2b6a9e5a2535295d2de03e27c2438d1f/src/wallet/test/wallet_tests.cpp#L158

ryanofsky reviewed (approved):
utACK aaaa57c2aad3a6278559ae3db5d953c847c6ffcb. The change seems fine, though it's not really clear what the motivation for it is. It might be good to add a release note describing the change and what benefits it has from a user perspective.

jamesob commented:
utACK https://github.com/bitcoin/bitcoin/pull/15870/commits/aaaa57c2aad3a6278559ae3db5d953c847c6ffcb

Makes a lot of sense to allow rescanning for as long as we can as @gmaxwell notes above - and indeed it might even be prudent for most users to set a future-proofing prune value that reflects their disk limitations at startup time. Better that your node start pruning vs. run out of disk space and exit. 

Maybe in the future we could even default to such a thing based on the datadir device if pruning-enabled-but-not-done-yet behavior is equivalent to a non-pruning node. Or introduce an "adaptive" pruning mode, which periodically polls for available disk space and adjusts the prune target accordingly. But of course this is all out of scope for this change.

promag commented:
@jamesob you mean an option to ensure a minimum free space?

Commit:
[doc] rpcwallet: Only fail rescan when blocks have been pruned

MarcoFalke commented on src/wallet/rpcdump.cpp:
```diff
@@ -573,7 +582,10 @@ UniValue importwallet(const JSONRPCRequest& request)
+        // Exit early and print an error.
+        // If a block is pruned after this check, we will import the key(s),
+        // but fail the rescan with a generic error.
         throw JSONRPCError(RPC_WALLET_ERROR, "Importing wallets is disabled in pruned mode");
```
Done (and added a release note)

promag commented on src/wallet/wallettool.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2016-2018 The Bitcoin Core developers
+// Copyright (c) 2016 The Bitcoin Core developers
```
aaaa57c2aad3a6278559ae3db5d953c847c6ffcb

👀

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -573,8 +582,11 @@ UniValue importwallet(const JSONRPCRequest& request)
+        // Exit early and print an error.
+        // If a block is pruned after this check, we will import the key(s),
+        // but fail the rescan with a generic error.
+        throw JSONRPCError(RPC_WALLET_ERROR, "Importing wallets is disabled when blocks are pruned");
```
fa7e311e169349bfcf1dab8b980724e8ddf4e749

This should be in 1s commit.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -573,7 +582,10 @@ UniValue importwallet(const JSONRPCRequest& request)
             }.ToString());
 
-    if (pwallet->chain().getPruneMode()) {
+    if (pwallet->chain().havePruned()) {
```
faf3729242c5e9486e137e549f05bf20bd3908a0

The dump file contains timestamps so, correct me if I'm wrong, it could only throw this error if earliest is pruned - the same for `importmulti` if timestamps are given.

MarcoFalke replied:
Sounds like a nice follow up pull request

promag reviewed (commented):
This PR would be just fine without aaaa57c2aad3a6278559ae3db5d953c847c6ffcb IMO.

utACK fa7e311e169349bfcf1dab8b980724e8ddf4e749.

jamesob commented:
> @jamesob you mean an option to ensure a minimum free space?

If you mean "prune as little as possible and only until necessitated by lack of space," yes. Some kind of adaptive prune would just be a measure to avoid hitting an out-of-disk error for as long as possible. But this is out of scope in terms of this PR so maybe I'll file an issue later or something.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -325,7 +325,11 @@ class ChainImpl : public Chain
-    bool getPruneMode() override { return ::fPruneMode; }
+    bool havePruned() override
+    {
+        LOCK(cs_main);
```
I'm not suggesting you change this PR, but for future reference we now have a handy `WITH_LOCK` macro: https://github.com/bitcoin/bitcoin/blob/fd61b9fc22faaa946ca43211c6c43b40ca4a860c/src/sync.h#L209

MarcoFalke replied:
I fail to see how this makes the code any easier to read

jnewbery reviewed (commented):
utACK fa7e311e169349bfcf1dab8b980724e8ddf4e749

Sjors commented:
post-merge utACK
